Scripts/Spells: Ferocious bite implemented

matanshukry opened the pull request:
**Changes proposed:**

- Implemented Ferocious bite

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:**

**Tests performed:**

(Does it build, tested in-game, etc.)

- [x] It builds
- [x] Tested in-game

**Known issues and TODO list:** (add/remove lines as needed)

**Dependant On:**

#26097

matanshukry committed:
POWER_BURN should not cancel when target isn't caster

matanshukry committed:
Merge branch 'master' into issue/ferocious-bite

matanshukry committed:
Added script for ferocious bite

matanshukry committed:
Merge remote-tracking branch 'origin/master' into issue/ferocious-bite

matanshukry committed:
Removed unused variable

illfated commented:
Utterly and purely cosmetic, but for the record: "Ferocious Bite" is the name of the spell, so you could choose to let "implemented" or "implement" be the only non-capitalized word in the title.

Shauren commented on src/server/scripts/Spells/spell_druid.cpp:
```diff
@@ -1510,6 +1510,40 @@ class spell_dru_wild_growth : public SpellScriptLoader
+
+    void Register() override
+    {
+        OnEffectHitTarget += SpellEffectFn(spell_dru_ferocious_bite::HandleHitTarget, EFFECT_1, SPELL_EFFECT_POWER_BURN);
```
This script is incrrect: EFFECT_1 (power burn) hits the caster so `GetHitDamage()` and `SetHitDamage()` are damage values dealt to caster from burned power

matanshukry replied:
Thanks for the thorough explaination! :)

Shauren commented on src/server/game/Spells/Spell.cpp:
```diff
@@ -5380,7 +5380,15 @@ SpellCastResult Spell::CheckCast(bool strict, uint32* param1 /*= nullptr*/, uint
             {
                 // Can be area effect, Check only for players and not check if target - caster (spell can have multiply drain/burn effects)
-                if (m_caster->GetTypeId() == TYPEID_PLAYER)
+                // effect target can be caster and can be different than spell target (e.g. Ferocious Bite)
```
Don't you have a separate PR for this part?

matanshukry replied:
I do, but this PR depends on it.
That's wyh I added the "Dependant On:" part in the OP and linked the other issue.

Shauren reviewed (changes requested):
Required hooks without hacking up CheckCast
`OnEffectLaunchTarget` for SPELL_EFFECT_POWER_BURN for caster - manually apply EFFECT_1 value from 102543 to GetEffectValue() (up to this point in calculation this will be your "max extra consumed power"). Damage multiplier will be `1.0f + std::min<float>(GetPower(POWER_ENERGY), maxConsumedExtraEnergy) / maxConsumedExtraEnergy`  and store this calculated value as a script class member (float _damageMultiplier). You do need make it "as if" ModifyPower was used without using it, we need it to actually use default effect handler later for proper packet contents

`OnEffectHitTarget` for SPELL_EFFECT_POWER_BURN for caster - manually apply EFFECT_1 value from 102543 to GetEffectValue()/SetEffectValue() and let it do its default thing later

`OnEffectHitTarget` for SPELL_EFFECT_SCHOOL_DAMAGE applying the multiplier saved from `OnEffectLaunchTarget` (with SetHitDamage)

matanshukry committed:
Completely removed extra condition, not needed

matanshukry committed:
Merge branch 'master' into issue/ferocious-bite

matanshukry committed:
Merge branch 'master' into issue/power_burn

matanshukry committed:
- Removed the entire switch case
- returned the ferocious bite code. Will be removed in another pr

matanshukry committed:
Merge branch 'master' into issue/ferocious-bite

matanshukry committed:
Merge branch 'issue/power_burn' into issue/ferocious-bite

matanshukry committed:
Merge branch 'master' into issue/ferocious-bite

matanshukry committed:
- Removed explicit care of function in EffectSchoolDMG
- Sorted class name and registration
- Handling the effect differently according to CR

matanshukry committed:
Commented unused argument

Shauren committed:
Merge branch 'master' into issue/ferocious-bite

Shauren committed:
rename sql

Shauren committed:
initialize variable

Shauren committed:
typo
